tests/src/providers/testqgswmsprovider.cpp

```diff
@@ -243,6 +247,52 @@ class TestQgsWmsProvider: public QObject
       QVERIFY( provider.layerMetadata().rights().at( 0 ).startsWith( "Base map and data from OpenStreetMap and OpenStreetMap Foundation" ) );
     }
 
+    void testConvertToValue()
+    {
+      QString dataDir( TEST_DATA_DIR );
+
+      QgsXyzConnection xyzConn;
+      xyzConn.url = QUrl::fromLocalFile( dataDir + QStringLiteral( "/maptiler_terrain_rgb_1.png" ) ).toString();
+      QgsRasterLayer layer( xyzConn.encodedUri(), "terrain", "wms" );
+      QVERIFY( layer.isValid() );
+      QVERIFY( layer.dataProvider()->dataType( 1 ) == Qgis::DataType::ARGB32 );
+
+      xyzConn.encodingSchemeKey = QStringLiteral( "maptilerterrain" );
+      QgsRasterLayer layer2( xyzConn.encodedUri(), "terrain", "wms" );
+      QVERIFY( layer2.isValid() );
+      QVERIFY( layer2.dataProvider()->dataType( 1 ) == Qgis::DataType::Float32 );
+
+      QgsMapSettings mapSettings;
+      mapSettings.setLayers( QList<QgsMapLayer *>() << &layer2 );
+      mapSettings.setExtent( layer2.extent() );
+      mapSettings.setOutputSize( QSize( 400, 400 ) );
+      mapSettings.setOutputDpi( 96 );
+      mapSettings.setDpiTarget( 48 );
+      QVERIFY( imageCheck( "convert_value_1", mapSettings ) );
```

Comment:
There are some strange visual artifacts in the converted image, for example in Antarctica or in Greenland. I am afraid it could be due to the fact that QGIS first does some interpolation in the RGB space, and only then we convert the interpolated values to floats, which may be giving those artifacts. If that's the case, we'll need to think how to fix it - either do the conversion before the interpolation or avoid the interpolation altogether if conversion is enabled.